build-system/tasks/build-bento.js

```diff
@@ -0,0 +1,271 @@
+  }
+  return componentsToBuild;
+}
+
+/**
+ * Watches for non-JS changes within an extensions directory to trigger
+ * recompilation.
+ *
+ * @param {string} compoentsDir
+ * @param {string} name
+ * @param {string} version
+ * @param {string} latestVersion
+ * @param {boolean} hasCss
+ * @param {?Object} options
+ * @return {Promise<void>}
+ */
+async function watchComponent(
+  compoentsDir,
+  name,
+  version,
+  latestVersion,
+  hasCss,
+  options
+) {
+  /**
+   * Steps to run when a watched file is modified.
+   */
+  function watchFunc() {
+    buildComponent(name, version, latestVersion, hasCss, {
+      ...options,
+      continueOnError: true,
+      isRebuild: true,
+      watch: false,
+    });
+  }
+
+  const cssDeps = `${compoentsDir}/**/*.css`;
+  const jisonDeps = `${compoentsDir}/**/*.jison`;
+  const ignored = /dist/; //should not watch npm dist folders.
+  watch([cssDeps, jisonDeps], {ignored}).on(
+    'change',
+    debounce(watchFunc, watchDebounceDelay)
+  );
+}
+
+/**
+ * Copies components from
+ * src/bento/components/$name/$name.js
+ * to
+ * dist/v0/$name-$version.js
+ *
+ * Optionally copies the CSS at components/$name/$version/$name.css into
+ * a generated JS file that can be required from the components as
+ * `import {CSS} from '../../../build/$name-0.1.css';`
+ *
+ * @param {string} name Name of the extension. Must be the sub directory in
+ *     the components directory and the name of the JS and optional CSS file.
+ * @param {string} version Version of the extension. Must be identical to
+ *     the sub directory inside the extension directory
+ * @param {string} latestVersion Latest version of the extension.
+ * @param {boolean} hasCss Whether there is a CSS file for this extension.
+ * @param {?Object} options
+ * @param {!Array=} extraGlobs
+ * @return {!Promise<void>}
+ */
+async function buildComponent(
+  name,
+  version,
+  latestVersion,
+  hasCss,
+  options,
+  extraGlobs
+) {
+  options = options || {};
+  options.extraGlobs = extraGlobs;
+  if (options.compileOnlyCss && !hasCss) {
+    return;
+  }
+  const componentsDir = `src/bento/components/${name}/${version}`;
+  if (options.watch) {
+    await watchComponent(
+      componentsDir,
+      name,
+      version,
+      latestVersion,
+      hasCss,
+      options
+    );
+  }
+
+  if (hasCss) {
+    if (!existsSync('build')) {
+      mkdirSync('build');
+    }
+    if (!existsSync('build/css')) {
+      mkdirSync('build/css');
+    }
+    await buildExtensionCss(componentsDir, name, version, options);
+    if (options.compileOnlyCss) {
+      return;
+    }
+  }
+  if (options.npm) {
+    await buildNpmBinaries(componentsDir, name, options);
+    await buildNpmCss(componentsDir, options);
+  }
+  if (options.binaries) {
+    await buildBinaries(componentsDir, options.binaries, options);
+  }
+  if (options.isRebuild) {
+    return;
+  }
+
+  const bentoName = getBentoName(name);
+  await buildExtensionJs(componentsDir, bentoName, {
```

Comment:
This function conditionally kicks off a decent number of async tasks and awaits each one in sequence. If they aren't order-dependent, it might make sense to create a `tasks = []` list and directly append each promise to it in the loop, then `return Promise.all(tasks)`.

Reply:
TBH this whole file is mostly a tweaked version of `build-system/tasks/extension-helpers.js` and I was attempting to change as little as possible in this initial version. There is a lot strange kinda funky code there that I would like to clean up in a future iteration.